### PR TITLE
Created combat controllers for different enemy types

### PR DESCRIPTION
**Pull request by trudeaua**
- Added combat controllers for bosses and enemies with one, two, and three attacks
-  Added 10 enemy prefabs
    - Created first boss (Big bad crab asset) named **Kaa**
    - Came up with names for the other Chaocytes
- Synced enemy animations with their attacks
- Player weapons now glow, and the weapon colour matches their character colour

### CODE REVIEW

**Comment by trudeaua** [comment]
Could you guys review my changes once more? just added some stuff for spawn animations, adjusted combat cooldowns, enemy healthbars, and combat VFX character colour matching.

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/EnemyBrainController.cs [hunk 1]
```diff
@@ -121,7 +121,20 @@ private void MakeCombatDecision()
...
+                case 0:
+                    _combat.PrimaryAttack();
+                    break;
+                case 1:
+                    _combat.SecondaryAttack();
+                    break;
+                case 2:
+                    _combat.Spell();
+                    break;
+            }
```
Comment (camvr) [inline Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/EnemyBrainController.cs]: We'll have to change this sometime soon, but I think it can wait until after implementation 1. Maybe open an issue assigned to you and I?

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/EnemyCombatController.cs [hunk 2]
```diff
@@ -2,30 +2,39 @@
...
+    public GameObject spawnVFX;
+
     protected List<GameObject> Players;
 
     private void Start()
     {
         Players = PlayerManager.Players;
+        StartCoroutine(CreateVFX(spawnVFX, new Vector3(gameObject.transform.position.x, 0.1f, gameObject.transform.position.z), 
+            Quaternion.identity, pl.GetColour(Stats.characterColour), 0.5f));
+        StartCoroutine(Spawn(gameObject, 0.05f, 0.9f));
```
Comment (camvr) [inline Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/EnemyCombatController.cs]: What is going on here?

Comment (camvr) [reply Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/EnemyCombatController.cs]: I think this is how you're doing weapon colours or something? Maybe add a comment to this part

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/EnemyCombatController.cs [hunk 3]
```diff
@@ -34,7 +43,41 @@ public virtual void PrimaryAttack()
...
+
+        Anim.SetTrigger("Spell");
+
+        // Attack any enemies within the attack sweep and range
+        foreach (GameObject player in Players.Where(player => CanDamageTarget(player.transform.position, attackRadius, attackAngle)))
+        {
+            // Calculate and perform damage
+            StartCoroutine(PerformDamage(player.GetComponent<EntityStatsController>(), Stats.ComputeDamageModifer(), 0.6f));
+        }
+    }
```
Comment (camvr) [inline Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/EnemyCombatController.cs]: Not a big fan of having such a specific function for this here... I would suggest sticking to the `primaryAttack` and `secondaryAttack` generalized scheme, then allow specific subclasses to implement this instead (rather than having an arbitrary `spell` function that only a few enemy controllers actually use

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/EntityCombatController.cs [hunk 4]
```diff
@@ -10,7 +10,9 @@ public class EntityCombatController : MonoBehaviour
     protected float AttackCooldown;
 
     protected Collider[] Hits = new Collider[32];
-    
+    protected PlayerColours pl = new PlayerColours();
```
Comment (camvr) [inline Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/EntityCombatController.cs]: Why is this here? I had this in the PlayerManager component because it took advantage of its singleton pattern (i.e. there should only be 1 instance of this, otherwise changes colours in the inspector would be a massive pain). I believe before it could be accessed using `PlayerManager.PlayerColours` or something similar

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/EntityStatsController.cs [hunk 5]
```diff
@@ -54,14 +44,15 @@ protected virtual void Update()
     {
         if (!isDead)
             health.Regen();
+        Debug.Log(health.CurrentValue);
```
Comment (camvr) [inline Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/EntityStatsController.cs]: Leftover Debug.Log?

Comment (camvr) [inline Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/EntityStatsController.cs]: Leftover Debug.Log?

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Player/Combat Classes/PlayerMeleeCombatController.cs [hunk 6]
```diff
@@ -75,25 +77,29 @@ protected override void OnPrimaryAttack(InputValue value)
...
             }
         }
     }
+
+    private IEnumerator FinishPrimaryAttack()
+    {
+        yield return new WaitForSeconds(primaryAttackCooldown);
+        Debug.Log(primaryAttackCooldown);
+        Anim.SetBool("PrimaryAttack", false);
+    }
```
Comment (camvr) [inline Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Player/Combat Classes/PlayerMeleeCombatController.cs]: Noticed this pattern in a few other functions... Perhaps there's a better, more extendable way to do this? We can discuss this later though

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Game/PlayerManager.cs [hunk 7]
```diff
@@ -42,36 +42,63 @@ private void Awake()
...
     }
 
     public static void RegisterPlayer(GameObject player)
     {
         Players.Add(player);
+
+        // assign the player a colour as soon as they're registered
+        CharacterColour characterColour = availableColours[Random.Range(0, availableColours.Count)];
+        availableColours.Remove(characterColour);
+        player.GetComponent<EntityStatsController>().characterColour = characterColour;
```
Comment (camvr) [inline Cosmic Quest - Order and Chaos/Assets/Scripts/Game/PlayerManager.cs]: Lol I like this alot actually

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Game/PlayerManager.cs [hunk 8]
```diff
@@ -42,36 +42,63 @@ private void Awake()
...
+                foreach (Material m in weaponMaterials)
+                {
+                    weaponMaterials[i].EnableKeyword("_EMISSION");
+                    weaponMaterials[i].SetColor("_Color", color);
+                    weaponMaterials[i].SetColor("_EmissionColor", color * intensity);
+                    i++;
+                }
+            }
+        }
+    }
```
Comment (camvr) [inline Cosmic Quest - Order and Chaos/Assets/Scripts/Game/PlayerManager.cs]: To be honest, this really doesn't belong here... The PlayerManager ultimately is going to be the thing that assigns the player characters to their gamepads and assigns them a colour, and other high level tasks. This is getting abit too specific... I think this should go in the one of the player components (I've been chucking most things into the StatsController since its kinda the "top level" player controller). When gamepad assignment is implemented I will also be moving the outline material generation to the player stats anyways (should be handled by the player controller itself to set the player to its assigned colours, the PlayerManager should only assign it the colour... Less stuff to worry about implementing in a scene this way).

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Items/Projectile.cs [hunk 9]
```diff
@@ -35,7 +35,7 @@ public void Launch(EntityStatsController launcherStats, Vector3 direction, float
         
         // Set position just in front of launcher
         _initialPosition = launcherStats.transform.position + launcherStats.transform.forward;
-        _initialPosition.y = ProjectileHeight;
+        _initialPosition.y = ProjectileHeight + 0.1f;
```
Comment (camvr) [inline Cosmic Quest - Order and Chaos/Assets/Scripts/Items/Projectile.cs]: Whats the point of this? Can't we change the value of ProjectileHeight instead?

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Items/Weapon.cs [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        
+    }
+
+    // Update is called once per frame
+    void Update()
+    {
+        
+    }
+}
```
Comment (camvr) [inline Cosmic Quest - Order and Chaos/Assets/Scripts/Items/Weapon.cs]: Can we just get rid of this file? The item system (I'm currently working on it) utilizes scriptable objects instead and this file would just get removed anyways

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/Combat Classes/EnemyTriAttackCombatController.cs [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+        foreach (GameObject player in Players.Where(player => CanDamageTarget(player.transform.position, attackRadius, attackAngle)))
+        {
+            // Calculate and perform damage
+            StartCoroutine(PerformDamage(player.GetComponent<EntityStatsController>(), Stats.ComputeDamageModifer(), secondaryAttackDelay));
+        }
+
+        AttackCooldown = secondaryAttackCooldown;
+    }
+
+    public override void Spell()
```
Comment (camvr) [inline Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/Combat Classes/EnemyTriAttackCombatController.cs]: As I stated in another comment, maybe we should treat this more as `tertiaryAttack`?

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/Combat Classes/EnemySingleAttackCombatController.cs [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+    public override void SecondaryAttack()
+    {
+        PrimaryAttack();
+    }
+
+    public override void Spell()
+    {
+        PrimaryAttack();
+    }
```
Comment (camvr) [inline Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/Combat Classes/EnemySingleAttackCombatController.cs]: These are unneccessary

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/EnemyCombatController.cs (hunk 2 above)
Comment (trudeaua) [reply Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/EnemyCombatController.cs]: Yep I created a new CreateVFX() function that takes in a colour and turns the colour of the VFX prefab into one that matches the player's colour! I'll add a comment

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/EntityCombatController.cs (hunk 4 above)
Comment (trudeaua) [reply Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/EntityCombatController.cs]: Alright, will make that change

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/EnemyCombatController.cs (hunk 3 above)
Comment (trudeaua) [reply Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/EnemyCombatController.cs]: Makes sense... seems obvious now that you pointed it out haha

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/EntityStatsController.cs (hunk 5 above)
Comment (trudeaua) [reply Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/EntityStatsController.cs]: Whoops!

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Game/PlayerManager.cs (hunk 8 above)
Comment (trudeaua) [reply Cosmic Quest - Order and Chaos/Assets/Scripts/Game/PlayerManager.cs]: OK, I'll try to get it working in the Stats controller

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Items/Projectile.cs (hunk 9 above)
Comment (trudeaua) [reply Cosmic Quest - Order and Chaos/Assets/Scripts/Items/Projectile.cs]: There was some weird issue with the projectile firing too low, I guess changing the projectile height should fix it

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Items/Weapon.cs (hunk 10 above)
Comment (trudeaua) [reply Cosmic Quest - Order and Chaos/Assets/Scripts/Items/Weapon.cs]: I can remove this, but then the player's weapons will not glow. I suppose I could find the objects by tag for now, in AssignWeaponColours()

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/Combat Classes/EnemyTriAttackCombatController.cs (hunk 11 above)
Comment (trudeaua) [reply Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/Combat Classes/EnemyTriAttackCombatController.cs]: Yeah good idea, I'll replace spell with tertiary attack

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Player/Combat Classes/PlayerMeleeCombatController.cs (hunk 6 above)
Comment (trudeaua) [reply Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Player/Combat Classes/PlayerMeleeCombatController.cs]: Yeah there was weird stuff happening when I was holding the trigger down for the healer and melee classes, because the AC bool would only be set to false when I let go of the trigger. So i made this function to set the bool to false when the animation finished. It works quite perfectly, but yeah there's probably a better way to do it.

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/EnemyCombatController.cs (hunk 2 above)
Comment (trudeaua) [reply Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/EnemyCombatController.cs]: So to clarify, no it's not doing weapon colours, but the colour of the spawn visual effect. I added a comment

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/EnemyCombatController.cs (hunk 3 above)
Comment (trudeaua) [reply Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/EnemyCombatController.cs]: Fixed

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/EntityCombatController.cs (hunk 4 above)
Comment (trudeaua) [reply Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/EntityCombatController.cs]: Fixed

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/EntityStatsController.cs (hunk 5 above)
Comment (trudeaua) [reply Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/EntityStatsController.cs]: Fixed

Comment (trudeaua) [reply Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/EntityStatsController.cs]: Fixed

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Game/PlayerManager.cs (hunk 8 above)
Comment (trudeaua) [reply Cosmic Quest - Order and Chaos/Assets/Scripts/Game/PlayerManager.cs]: Fixed

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Items/Projectile.cs (hunk 9 above)
Comment (trudeaua) [reply Cosmic Quest - Order and Chaos/Assets/Scripts/Items/Projectile.cs]: Fixed

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Items/Weapon.cs (hunk 10 above)
Comment (trudeaua) [reply Cosmic Quest - Order and Chaos/Assets/Scripts/Items/Weapon.cs]: Fixed

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/Combat Classes/EnemyTriAttackCombatController.cs (hunk 11 above)
Comment (trudeaua) [reply Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/Combat Classes/EnemyTriAttackCombatController.cs]: Fixed

File: Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/Combat Classes/EnemySingleAttackCombatController.cs (hunk 12 above)
Comment (trudeaua) [reply Cosmic Quest - Order and Chaos/Assets/Scripts/Entity/Enemy/Combat Classes/EnemySingleAttackCombatController.cs]: Fixed
